editor/src/clj/editor/grid.clj

```diff
@@ -13,192 +13,385 @@
+        label (doto (Label. "Opacity")
+                (.setPrefWidth 70))]
+    (ui/observe
+     (.valueProperty slider)
+     (fn [_observable _old-val new-val]
+       (let [val (math/round-with-precision new-val 0.01)]
+         (prefs/set! prefs prefs-path val)
+         (invalidate-grids! app-view))))
+    [label slider]))
+
+(defn plane-toggle-button
+  [prefs plane-group prefs-path plane]
+  (let [active-plane (prefs/get prefs prefs-path)]
+    (doto (ToggleButton. (string/upper-case (name plane)))
+      (.setToggleGroup plane-group)
+      (.setSelected (= plane active-plane))
+      (ui/add-style! "plane-toggle"))))
+
+(defmethod settings-row :active-plane
+  [app-view prefs option]
+  (let [prefs-path (conj grid-prefs-path option)
+        plane-group (ToggleGroup.)
+        buttons (mapv (partial plane-toggle-button prefs plane-group prefs-path) axes)
+        label (doto (Label. "Plane")
+                (HBox/setHgrow Priority/ALWAYS)
+                (.setPrefWidth 70))]
+    (ui/observe (.selectedToggleProperty plane-group)
+                (fn [_ ^ToggleButton old-value ^ToggleButton new-value]
+                  (if new-value
+                    (do (let [active-plane (-> (.getText new-value) string/lower-case keyword)]
+                          (prefs/set! prefs prefs-path active-plane))
+                        (invalidate-grids! app-view))
+                    (.setSelected old-value true))))
+    (concat [label] buttons)))
+
+(defmethod settings-row :color
+  [app-view prefs option]
+  (let [prefs-path (conj grid-prefs-path option)
+        text-field (TextField.)
+        [r g b a] (prefs/get prefs prefs-path)
+        color (->> (Color. r g b a) (.toString) nnext (drop-last 2) (apply str "#"))
+        label (doto (Label. "Color")
+                (HBox/setHgrow Priority/ALWAYS)
+                (.setPrefWidth 83))
+        cancel-fn (fn [_] (ui/text! text-field color))
+        update-fn (fn [_] (try
+                            (if-let [value (some-> (.getText text-field) colors/hex-color->color)]
+                              (do (prefs/set! prefs prefs-path value)
+                                  (invalidate-grids! app-view))
+                              (cancel-fn nil))
+                            (catch Exception _e
+                              (cancel-fn nil))))]
+    (doto text-field
+      (ui/text! color)
+      (ui/on-action! update-fn)
+      (ui/on-cancel! cancel-fn)
+      (ui/auto-commit! update-fn))
+    [label text-field]))
+
+(defn- axis-group
+  [app-view prefs prefs-path axis]
+  (let [text-field (TextField.)
+        label (Label. (string/upper-case (name axis)))
+        size-val (str (get (prefs/get prefs prefs-path) axis))
+        cancel-fn (fn [_] (ui/text! text-field size-val))
+        update-fn (fn [_] (try
+                            (let [value (some-> (.getText text-field) Float/parseFloat)]
+                              (if (pos? value)
```

Comment:
I get a boxed math warning from this line. I don't think the `some->` clause above is warranted, since `.getText` should always return a string, and `Float/parseFloat` will in fact throw an exception if the text can't be interpreted as a `float`. 🤔